README.md
CSS: /css/default.css

Artificial Intelligent
======================

This is my page(s) at GitHub.com. Integrated into the system's 
[jekyll](http://github.com/mojombo/jekyll) engine it's the best 
manifestation of my tremendous web-designing talent!

Used technics: markdown, liquid, jekyll, xhtml.

----
Author: Pavel _argent_smith_ Argentov <argentoff@gmail.com>


CHANGELOG
[6354bac | Wed Oct 27 14:26:57 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Finally updated CHANGELOG at v0.2.2

[3b717b2 | Wed Oct 27 14:22:23 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * v0.2.2 prerelease

[61a2ad2 | Wed Oct 27 14:19:40 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Version bump to 0.2.2

[d7affc1 | Wed Oct 27 14:10:11 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Added CHANGELOG

[61c87e1 | Wed Oct 27 13:22:40 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Working on documenation

    Started to standartize project's Git commit messages; added 'changelog' target
    to Rakefile; removed rubyforge references from Rakefile.

[2f262b0 | Wed Oct 27 13:22:02 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Version bump to 0.2.1

[86e8899 | Wed Oct 27 11:30:13 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * execute -> execute!

    In Rsh class: changed and renamed "execute" method though leaving it
    compatible with the previous version. Now execute! may be called with a remote
    command String as a parameter: it will be recorded in @command instance
    variable and then executed as usual.

[3a7b464 | Wed Oct 27 11:28:13 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Version bump to 0.2.0

[b937438 | Wed Oct 27 08:43:12 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Farewell, Rubyforge!

    Removed rubyforge_project from specification: it points to a wrong place (a
    completely different project with the same name). This project at the moment
    is hosted solely on GitHub.com; gem is published on rubygems.org.

[8d1072e | Wed Oct 27 08:38:42 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Version bump to 0.1.2

[2fe593f | Mon Oct 25 14:12:52 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Fixed some ugly rdoc problems; version bump 0.1.0 -> 0.1.1

[117da5b | Mon Oct 25 13:13:05 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Version bump to 0.1.1

[373016c | Sat Oct 23 04:57:18 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * regenerated gemspec

[a867003 | Sat Oct 23 04:55:46 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * minor changes to gemspec

[337714e | Fri Oct 22 15:22:52 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Finished the code: it seems to be working.

    Bumped VERSION to 0.1.0

[bbd9901 | Fri Oct 22 13:44:09 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Version bump to 0.1.0

[3a34fc9 | Thu Oct 21 15:43:28 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * slight rework on constructor test; added 'command' attribute

[f76ece2 | Thu Oct 21 15:28:35 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * added gemspec (rake generated)

[475091d | Thu Oct 21 15:26:41 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Trying to get the very idea of gem, rspec, etc.

[64289f3 | Thu Oct 21 12:19:07 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Version bump to 0.0.0

[77c129a | Thu Oct 21 12:00:49 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Struggling with gem specs...

[3fdca36 | Thu Oct 21 11:52:52 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Corrected some vars in Rakefile

[79d657e | Thu Oct 21 10:53:37 UTC 2010] Pavel Argentov <argentoff@gmail.com>

 * Initial commit to rsh.



spec/rsh_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')

describe "Rsh" do
    it "creates rsh instance" do
      rsh = Rsh.new
    end
end


spec/spec_helper.rb
$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
require 'rsh'
require 'spec'
require 'spec/autorun'

Spec::Runner.configure do |config|
  
end


lib/rsh_p_r.rb
#= Pure ruby RSH client implementation
#
# This implementation's idea is taken from Perl implementation, Net::Rsh
# by Oleg Prokopyev, <riiki@gu.net>
module RshPR
  require 'socket'
  # force pure ruby rsh client
  attr_accessor :ruby_impl
  # Local user name, used in pure ruby implementation.
  attr_accessor :luser

  # Specific constructor: additional parmeter key added:
  # * :ruby_impl -- turns on pure ruby implementation of RSH protocol client.
  def initialize(args={})
    args = {
      :host      => "localhost",
      :command   => "",
      :ruser     => "nobody",
      :luser     => ENV["USER"] || "nobody",
      :to        => 3,
      :nullr     => false,
      :ruby_impl => false
    }.merge!(args)

    @result     = ""
    @ruby_impl  = args[:ruby_impl]
    @host       = args[:host]
    @command    = args[:command]
    @ruser      = args[:ruser]
    @to         = args[:to]
    @nullr      = args[:nullr]

    begin
      if @ruby_impl
        @executable = :ruby_impl
      else
        open("| which rsh") do |io|
          @executable = io.gets.chomp
        end
      end
    rescue => detail
      @executable = :ruby_impl
    end
  end

  def execute!(command=nil)
    @command = command if command
    @result = ""
    if @executable == :ruby_impl
      # pure ruby implementation call
      rsh_ruby do |line|
        yield(line) if block_given?
        @result << line
      end
    else
      # OS'es rsh(1) call
      open "|#{@executable} #{"-n" if @nullr} -l#{@ruser} -t#{@to} #{@host} #{@command}" do |io|
        io.each do |line|
          yield(line) if block_given?
          @result << line
        end
      end
    end
    @result
  end

  private
  # pure ruby rsh client
  def rsh_ruby
    r = ""
    s = TCPSocket.new(@host, 514)
    s.write "0\0"
    s.write "#{@luser}\0"
    s.write "#{@ruser}\0"
    s.write "#{@command}\0"
    while line = s.gets
      if block_given?
        yield line
      else
        r << line
      end
    end
    s.close
    r == "" ? nil : r
  end
end

lib/rsh.rb
# -*- Ruby -*-
# Ruby remote shell protocol (RFC 1282) client.

require 'socket' # we'll need it in pure ruby implementation

# = Ruby remote shell protocol (RFC 1282) client
# 
# Creates and operates an rsh client instance. Parameters may
# be specified through either constructor or attribute accessors. Result of
# remote command execution (_String_) is either returned in functional style
# (<tt>execute!</tt> method call) or in special attribute, _:result_.
#
# == New: pure ruby rsh client
# As of ver. 1.0.0, pure ruby rsh client is implemented. It is used whether if
# you set :ruby_impl to true or if 'rsh' program wasn't found in the system while
# setting :ruby_impl to false.
#
# == Synopsis
#
#   require 'rsh'
#
#   rsh = Rsh.new(:host => "c7206", :ruser => "bill",
#                 :command => "show clock")
#
#   rsh.execute! do |line| puts line end
#
#   18:30:46.799 MSD Fri Oct 22 2010
#
# * See c) case in execute! method comment. 
#
# === Example:
#
# This is an example of rsh in action usin interactive Ruby.
#
#   irb(main):001:0> require 'rsh'
#   => true
#
#   irb(main):002:0> rsh = Rsh.new(:host => "c7206", :ruser => "bill",
#                                  :command => "show clock")
#   => #<Rsh:0x2853f390 @ruser="bill", @executable="/usr/bin/rsh", @result="", @command="show clock", @nullr=false, @host="c7206", @to=3>
#
#   irb(main):003:0> rsh.execute! do |line| puts line end
#
#   18:30:46.799 MSD Fri Oct 22 2010
#   => "\r\n18:30:46.799 MSD Fri Oct 22 2010\n"
#
#
# == See also
#
#   % man 1 rsh
#
class Rsh
  # flags whether to force pure ruby rsh client
  attr_accessor :ruby_impl
  # path to rsh program, +String+
  attr_reader   :executable
  # result +String+
  attr_reader   :result
  # remote server hostname or IP, +String+
  attr_accessor :host
  # remote command, +String+
  attr_accessor :command
  # remote username, +String+
  attr_accessor :ruser
  # Local user name, used in pure ruby implementation.
  attr_accessor :luser
  # rsh timeout, +Integer+ (see man 1 rsh)
  attr_accessor :to
  # boolean knob for <tt>/dev/null</tt> redirection; see man rsh for further
  # information
  attr_accessor :nullr

  # The Constructor. If :ruby_impl is false checks the presence of rsh in the system (running,
  # naturally, 'which rsh') and prepares the command to be run with <tt>execute!</tt>. 
  # rsh CLI arguments are either having default values, being collected from constructor 
  # call or specified via accessors.
  #
  # Call sequence:
  #   Rsh.new #=> Rsh instance
  #   Rsh.new(:host      => "hostname",
  #           :command   => "example.sh",
  #           :ruser     => "jack",
  #           :to        => 5,
  #           :nullr     => true
  #           :ruby_impl => true) #=> Rsh instance
  # 
  # If called without arguments, the following default values are being used:
  # 
  #   :host      => "localhost"
  #   :command   => ""
  #   :ruser     => "nobody"
  #   :to        => 3
  #   :nullr     => false
  #   :ruby_impl => false
  #   :luser     => ENV["USER"] || "nobody"
  #
  def initialize(args={})
    args = {
      :host      => "localhost",
      :command   => "",
      :ruser     => "nobody",
      :luser     => ENV["USER"] || "nobody",
      :to        => 3,
      :nullr     => false,
      :ruby_impl => false
    }.merge!(args)

    @result     = ""
    @ruby_impl  = args[:ruby_impl]
    @host       = args[:host]
    @command    = args[:command]
    @ruser      = args[:ruser]
    @luser      = args[:luser]
    @to         = args[:to]
    @nullr      = args[:nullr]

    begin
      open("| which rsh") do |io|
        @executable = io.gets.chomp
      end
    rescue => detail
      @ruby_impl = true
    end
  end
  
  # :ruby_impl writer ensuring that we can set/reset :ruby_impl only
  # if @executable is not nil.
  def ruby_impl=(bool)
    @executable ? @ruby_impl = bool : @ruby_impl
  end

  # Executes rsh command using previously collected arguments. If :ruby_impl is
  # true or system rsh binary was not found during initialize runs pure ruby rsh
  # client.
  #
  # If given a block, calls it for each line received from rsh output
  # (parameter _line_).
  #
  # Call sequence:
  #
  # a)
  #   rsh_inst.execute! do |line|
  #     puts line
  #   done #=> complete_result_String
  # b)
  #   rsh_inst.execute! #=> complete_result_String
  # c)
  #   rsh_inst.execute! "remote_command" #=> same as above, but replacing
  #                                          @command with the given string
  #
  # Returns:: the complete rsh output as one _String_. The result is also
  #           stored and available via _result_ attribute.
  #
  def execute!(command=nil)
    @command = command if command
    @result = ""
    if @ruby_impl
      # pure ruby implementation call
      rsh_ruby do |line|
        yield(line) if block_given?
        @result << line
      end
    else
      # OS'es rsh(1) call
      open "|#{@executable} #{"-n" if @nullr} -l#{@ruser} -t#{@to} #{@host} #{@command}" do |io|
        io.each do |line|
          yield(line) if block_given?
          @result << line
        end
      end
    end
    @result
  end

  private
  # pure ruby rsh client
  def rsh_ruby
    r = ""
    s = TCPSocket.new(@host, 514)
    s.write "0\0"
    s.write "#{@luser}\0"
    s.write "#{@ruser}\0"
    s.write "#{@command}\0"
    s.getc # cutting off the first byte, signalling?
    while line = s.gets
      if block_given?
        yield line
      else
        r << line
      end
    end
    s.close
    r == "" ? nil : r
  end

  alias :execute :execute!

end
